### ISSUE TITLE
Safer alternative to specifying mysqldump password on the command line for db:dump?

### ISSUE THREAD

**Issue by driskell** [netz98/n98-magerun#490]
I can see that magerun runs mysqldump during db:dump, and specifies the password on the command line. This is momentarily visible to all users on the system in process lists until mysqldump clears it out of memory. So although not end of the world, it is relatively insecure (running it like this interactively, mysqldump would actually warn about the insecure nature of the method.)

One proposal I can think is to create a temporary MySQL configuration file, for mysqldump to use, placed next to local.xml with the same permissions. This has "[client]" section and "password =". mysqldump command can then refer to it. If local.xml is not safe then this file wouldn't be too. I can see special care might need to be taken for this though.

Another proposal (separate, or in combination with) could be a "--no-password" option and then in combincation with --command-only a user can make their own decision on method.

What are people's thoughts on this?

**Comment by tkn98** [comment]
The different available options are outlined in the vendors documentation here:

- [End-User Guidelines for Password Security](http://dev.mysql.com/doc/refman/5.6/en/password-security-user.html)

I assume that the config-editor variant is not suiteable. Passing the password via switch will issue a warning with newer mysql versions (5.6+ IIRC).

**Comment by hostep** [comment]
Just tuning in to mention it would indeed be a good idea to use a temporary configuration file as you both suggest.
I noticed a bunch of warnings while using Mysql 5.6, especially when running the tests, which makes it a lot harder to interpret:
```
...........SS..................I................Warning: Using a password on the command line interface can be insecure.
.Warning: Using a password on the command line interface can be insecure.
.Warning: Using a password on the command line interface can be insecure.
.Warning: Using a password on the command line interface can be insecure.
.Warning: Using a password on the command line interface can be insecure.
.Warning: Using a password on the command line interface can be insecure.
.Warning: Using a password on the command line interface can be insecure.
.....Warning: Using a password on the command line interface can be insecure.
....  63 / 192 ( 32%)
............................................................... 126 / 192 ( 65%)
.SS..Warning: Using a password on the command line interface can be insecure.
.......................................................... 189 / 192 ( 98%)
...
```

So using a temp config file will remove these.

**Comment by tkn98** [comment]
Instead of a temporary config file, you can as well (and it's easier) set an environment variable. It can be insecure, too. Just saying. Writing a temp config file can be insecure, too.

**Comment by driskell** [comment]
Environment variables have the same problem as command line. They appear in process listing.

Writing a temp config file CAN be secure. And if you maintain the same permissions as the local.xml file - the security is guaranteed to be the same as that applied to local.xml. If local.xml can be read... well you're at least not making anything worse.

**Comment by tkn98** [comment]
Yes, environment variables have a similar problem. And writing to a temp file CAN be secure.

Question is if someone wants to put the extra effort into integrating with a tempfile at first just to get the warning out of the way.

With local.xml is a good point, that's why I wouldn't stress so much for the temp file but think it would be OK to use the environment variable.

However I don't know for sure which users can read others users environment variables from process listing (either directly or indirectly).

**Comment by driskell** [comment]
@tkn98 It's more a case with env vars, it's worse. With command line, mysqldump can (on most systems) rewrite the argv with asterisks so that once started up, the process listing shows asterisks instead of the password. But of course there's a race condition if you spam process list at backup time. With environment variable I believe there's no way for the process to rewrite it - so it's stuck in there, and most systems do allow listing of environment variables in process listing.

See https://dev.mysql.com/doc/refman/5.1/en/password-security-user.html, last point. MySQL consider password on the command line "insecure", but password in the ENV, "extremely insecure",

Regarding the temp file next to local.xml, I think it is by far the best solution, if care is taken to avoid symlink redirection vulnerabilities and the likes (though if those are possible, I think local.xml is vulnerable already). It is also the place where security CAN be controlled using file system permissions, in the SAME way we secure the local.xml which contains the password anyway. With command line and ENV the security is out of anyone's control - thus why MySQL request you avoid both methods.

I'm happy to try implementing this and propose the code changes if that was one of your questions?

**Comment by tkn98** [comment]
Just ran a quick test to have fatal-error safe tmpfiles (deleted also in case of PHP crash), and it works with `tmpfile()`:

	$temp = tmpfile();

filename is available via

	$meta_data = stream_get_meta_data($temp);
	echo $filename = $meta_data["uri"], "\n";

(tested on linux)

If you can file a patch, please do!

**Comment by driskell** [comment]
I'd be hesitant to store it away from local.xml (which gives some guarantee of security). But I guess if tmpfile() is inherently secure (seems it will be by default) then it may be even better - so that's a good idea.

I'm thinking an "--insecure" option though in case people have tmpfile problems.

**Comment by tkn98** [comment]
A switch sounds fair (perhaps not necessary for first implementation), just perhaps not name it "insecure" as unclear, just something like --use-passwort-switch or something into that direction. what's considered secure or insecure is likely to change over time.

**Comment by driskell** [comment]
Naming it insecure will ensure people are aware of the inherent risks, which I believe many many people will not be. If it turns out in the future the password switch does become secure, then the switch can be changed.

If it would get in the way of merge though I can hold back on that :-) But I feel it's so easy for people to share things like this and no inform people what is happening, and it ensures readers of any guides/articles will be encouraged to question and learn the security aspects.

I don't expect it the insecurity of --password or ENV etc will change too as it's deeply embedded in the way Linux works. The documentation I linked to by MySQL regarding insecure use of --password and environment variables has been in the MySQL documentation since 3.22 at least [1], released a whopping 16 or so years ago (documentation generated 1999). It's shocking to me the WARNING actually only started appearing in 2012 in 5.6.5 [2]. Clearly very few read the documentation before using things :)

1. http://www-sbras.nsc.ru/win/docs/db/mysql/manual.html
2. https://bugs.mysql.com/bug.php?id=66546

I'll whip something up though with tmpfile() and we can see how well it works, and move from there. We can easily get philosophical and at the end of day it's Collaborators like yourself, and Maintainers that need to be comfortable with the change as they'll receive the Issues - so I'll try keep it simple and let you do the rest :-)

**Comment by tkn98** [comment]
No don't get me wrong. Just create a patch as you seem fit, you don't have to argue for every detail you see fit, it was just what popped into my mind. I don't even meant that my suggestion was better, just different to highlight an idea which came to my mind (by making up some difference).

Even some name turns out to be wrong in the future, we can change that, so don't overvalvue the distinct feedback I was giving. Was not meant to say you should hold back with what you think would make magerun better.

And yes, warning is only since 5.6 - and not with the (more) insecure ENV as you pointed it out.

**Comment by driskell** [comment]
Ah sorry! :) I guess it's good to document all the ideas and +1/-1 etc. I'll try get something raised for discussion soon.

If anyone interested, at the moment I just cheeky wrap like this, and use my own secured "~/.my.cnf" defaults file for the user running it.

```
./n98-magerun.phar --root-dir=... db:dump --only-command --strip=... --compression=gzip magento.sql.gz | sed 's/-p[^ ][^ ]* //' | bash
```

**Comment by driskell** [comment]
@tkn98 I've whipped something up but come across some problems.
So the options I have are "--no-password" and "--insecure-password". Default is to use tmpfile() for Dump/Import/Query.

Info command I've left it as before (so it defaults to --insecure-password essentially). I've just added an options parameter to getMySQLClientToolConnectionString to facilitate this.

If the user specifies --only-command though... the tmpfile() is gone afterwards. So that's backwards compat possibly gone. What do you think here?

**Comment by tkn98** [comment]
It's hard to give some suggestions w/o seeing the code. If --only-command is used and the new option is in, sure, the tmpfile is gone. So perhaps it should add --insecure-password implicitly when used. So it's backward compatible (just if none of --no-password" and "--insecure-password" was given).

**Comment by amenk** [comment]
Please keep in mind that the user running magerun might not have the permission to write a temporary file in `$MAGENTROOT/app/etc/`

**Comment by orrd** [comment]
For MySQL >= 5.6, a good solution may be to use mysql_config_editor to set a "login-path" that stores the password in an encrypted form in the user's home directory.

http://stackoverflow.com/questions/20751352/suppress-warning-messages-using-mysql-from-within-terminal-but-password-written
